src/extensions/registries/page-registry.ts

```diff
@@ -1,93 +1,95 @@
-  routePath: string; // full route-path to registered extension page
+  params?: UrlParam[];
 }
 
 export interface PageComponents {
   Page: React.ComponentType<any>;
 }
 
-export function getExtensionPageUrl<P extends object>({ extensionId, pageId = "", params }: PageMenuTarget<P>): string {
-  const extensionBaseUrl = compile(`/extension/:name`)({
-    name: sanitizeExtensionName(extensionId), // compile only with extension-id first and define base path
-  });
-  const extPageRoutePath = path.posix.join(extensionBaseUrl, pageId);
+export interface PageTarget<P = {}> {
+  extensionId?: string;
+  pageId?: string;
+  params?: Record<string, any | any[]> & P; // default target page params
+}
+
+export interface RegisteredPage extends PageRegistration {
+  extensionId: string;
+  url: string; // registered extension's page URL (without page params)
+}
+
+export function getExtensionPageUrl(target: PageTarget): string {
+  const { extensionId, pageId = "", params: targetParams = {} } = target;
+  let stringifiedParams = "";
+
+  // stringify params to matched target page
+  const page = globalPageRegistry.getByPageTarget(target) || clusterPageRegistry.getByPageTarget(target);
+
+  if (page?.params) {
+    const searchParams = page.params.map(urlParam => {
+      return urlParam.toSearchString({
+        value: targetParams[urlParam.name] ?? urlParam.getDefaultValue(),
+        mergeGlobals: false,
+        withPrefix: false,
+      });
+    });
 
-  if (params) {
-    return compile(extPageRoutePath)(params); // might throw error when required params not passed
+    if (searchParams.length > 0) {
+      stringifiedParams = `?${searchParams.join("&")}`;
+    }
   }
 
-  return extPageRoutePath;
+  return path.posix.join("/extension", sanitizeExtensionName(extensionId), pageId, stringifiedParams);
 }
 
 export class PageRegistry extends BaseRegistry<RegisteredPage> {
   @action
-  add(items: PageRegistration | PageRegistration[], ext: LensExtension) {
-    const itemArray = rectify(items);
-    let registeredPages: RegisteredPage[] = [];
-
+  add(pages: PageRegistration | PageRegistration[], extension: LensExtension) {
     try {
-      registeredPages = itemArray.map(page => ({
-        ...page,
-        extensionId: ext.name,
-        routePath: getExtensionPageUrl({ extensionId: ext.name, pageId: page.id }),
-      }));
-    } catch (err) {
-      logger.error(`[EXTENSION]: page-registration failed`, {
-        items,
-        extension: ext,
-        error: String(err),
-      });
-    }
+      const items = [pages].flat().map(page => this.registerPage(page, extension));
 
-    return super.add(registeredPages);
-  }
-
-  getUrl<P extends object>({ extensionId, id: pageId }: RegisteredPage, params?: P) {
-    return getExtensionPageUrl({ extensionId, pageId, params });
+      return super.add(items);
+    } catch (error) {
+      return Function; // no-op
```

Comment:
why not `noop`? Then there wouldn't be need for the comment.
```suggestion
      return noop;
```